### ISSUE TITLE
using other (static) widgets

### ISSUE THREAD

**Issue by pterrien** [cozy-comfyui/cozy_ex_dynamic#3]
First, I would like to thank you for this very interesting example you provide us. I found another approach in a different repo, but yours is by far more elegant.

I face an issue when I adapt the node to have a static widget (defined in the node `INPUT_TYPES`), converted to an input slot:  when connecting a link, the static one gets renamed to either `_PREFIX + '_'` or `_PREFIX + '_1'` (depending if we connect the static or dynamic one). Which is normal, when analyzing the code.

The solution I found is to patch `nodeType.prototype.onConnectionsChange` with:
1. early exit if the slot name doesn't start with `_PREFIX`, something like (line 44):
```js
if (!node_slot.name.startsWith(_PREFIX)) return me;
```

2. Modify line 78 to have:
```js
slot.name = name === _PREFIX ? `${name}_${count}` : name;
```

Of course, using a regexp would be better than `startsWith()`, did that to make it more readable.

It seems to work, but I find it a bit dirty... would you have a cleaner approach?

Bonus question: is there a way to ensure the static input is always listed before the dynamic ones? I'm afraid it's part of Comfy default behavior and can't be changed easily though...

Thanks in advance for your help!

**Comment by pterrien** [comment]
While digging, I may have identified another issue (possibly a bug).
For my testing, I just changed `_PREFIX = 'str'` and `_TYPE = 'STRING'` (and of course the Python class accordingly) to make it easier, but it shouldn't change the behavior.
Let's imagine I have 4 string inputs, with values: 'input1', 'input2', 'input3', 'input4'

1. I first connect the first 3 inputs, so I now have:
  input1 -> str_1
  input2 -> str_2
  input3 -> str_3
  str (not connected)

2. Now, if I connect the 4th input to str_1, replacing the existing link, I would expect:
  input4 -> str_1
  input2 -> str_2
  input3 -> str_3
  str (not connected)
But instead I get:
  input4 -> str_1
  input3 -> str_2
  str (not connected)

So my `input2`, previously connected to `str_2`, is  also removed.

If I analyze correctly, the main issue is that replacing a link is not a single event, but 2 distinct ones: a disconnect then a connect. Possibly also the fact that `this.removeInput()` may get called twice, in case of a disconnect?

Do you see a way to correct this?

Thank you!

**Comment by Amorano** [comment]
For the static "entries" -- my own solution is to put those at the top and only put the dynamic fields after. Trying to mix them, or put the dynamic entries above static ones, only leads to a bunch of confusion and code work. That is, sadly, a problem for ComfyUI itself and how its rendering the inputs.

As for the second problem, correct, that is also a ComfyUI problem. This was, unfortunately, a simple example how to at least get a dynamic input, not a complete solution for all the setups that you can do/need with dynamic inputs.

I don't want to confuse this solution (simple) with more complex problems since most are only looking for an example how-to.

For the more complex problem(s) you are talking about, I would recommend looking at:

[my repo](https://github.com/Amorano/Jovimetrix) or
[comfy_mtb](https://github.com/melMass/comfy_mtb) or
[kjnodes](https://github.com/kijai/ComfyUI-KJNodes)

which have solved these issues more complicated issues with a little more code work around the problem of dynamic inputs.

Sorry! I hope that helps.

**Comment by pterrien** [comment]
Many thanks for this quick and detailed answer!
I fully understand that your repo is intended as an example and should  not be too complex, that's not its purpose.

In the meantime, I took some time to study and import some functions from the comfy_mtb repo, and had some success with it. It seems to handle my 2nd issue (by using an approach of "transit"), with the (acceptable) downside of not removing the last dyn inputs on manual disconnect (but replacing works).
And I also managed to patch it to handle static inputs, with the limitation you described about order of inputs.

All things considered, I'm now able to use dynamic inputs for my own personal nodes, which is great!
So I am really grateful for this great repo you provide us, indeed it did a great job for me to have a better understanding of dynamic inputs and build my own solution.

Thanks again, cheers!
